sqlmesh/core/macros.py

```diff
@@ -370,9 +390,30 @@ def add_one(evaluator: MacroEvaluator, column: exp.Literal) -> exp.Add:
     registry_name = "macros"
 
     def __call__(
-        self, func: t.Callable[..., DECORATOR_RETURN_TYPE]
+        self,
+        func: t.Callable[..., DECORATOR_RETURN_TYPE],
     ) -> t.Callable[..., DECORATOR_RETURN_TYPE]:
-        wrapper = super().__call__(func)
+        spec = inspect.getfullargspec(func)
+
+        @wraps(func)
+        def _typed_func(*args: t.Any, **kwargs: t.Any) -> DECORATOR_RETURN_TYPE:
+            """Coerce arguments where possible to the user-defined type annotations."""
+            evaluator: MacroEvaluator = args[0]
+            argmap = inspect.getcallargs(func, *args, **kwargs)
+            for param, value in argmap.items():
+                coercible_type = spec.annotations.get(param)
+                if not coercible_type:
+                    continue
+                if isinstance(coercible_type, str):
```

Comment:
what does it mean for the coercible type to be a string? why do you continue here? is this like 

def x(y: "my custom class") ?

Reply:
Ah its actually extremely interesting @tobymao. This specifically happens when you have `from __future__ import annotations` at the top of the file. The solution is to use `typing.get_type_hints`. 

See PEP 563 https://peps.python.org/pep-0563/

I was chasing this earlier today. Really nice to get to the bottom of it.

<img width="862" alt="image" src="https://github.com/TobikoData/sqlmesh/assets/41213451/7377f008-09cf-4d9c-badd-1499afa74eb6">

Reply:
I will update the code shortly, current code was just avoiding the problem since I hadnt fully gotten it yet.